crates/bevy_asset/src/temp.rs

```diff
@@ -0,0 +1,48 @@
+use std::path::{Path, PathBuf};
+
+use bevy_ecs::{system::Resource, world::World};
+
+use crate::io::AssetSourceBuilder;
+
+#[derive(Resource)]
+struct TempAssetDirectory {
+    folder: TempDirectory,
+}
+
+impl TempAssetDirectory {
+    fn path(&self) -> &Path {
+        self.folder.path()
+    }
+}
+
+enum TempDirectory {
+    Delete(tempfile::TempDir),
+    Persist(PathBuf),
+}
+
+impl TempDirectory {
+    fn path(&self) -> &Path {
+        match self {
+            TempDirectory::Delete(x) => x.path(),
+            TempDirectory::Persist(x) => x.as_ref(),
+        }
+    }
+}
+
+pub(crate) fn get_temp_source(
+    world: &mut World,
+    temporary_file_path: Option<String>,
+) -> AssetSourceBuilder {
+    let temp_dir = world.get_resource_or_insert_with::<TempAssetDirectory>(|| {
+        let folder = match temporary_file_path {
+            Some(path) => TempDirectory::Persist(path.into()),
+            None => TempDirectory::Delete(tempfile::tempdir().expect("todo")),
+        };
+
+        TempAssetDirectory { folder }
+    });
+
+    let path = temp_dir.path().to_str().expect("todo");
```

Comment:
TODO: Likewise, the `OsStr` to `str` conversion here can technically fail, but I think it's fair to assume it _will_ succeed. But, I still don't like it.

Reply:
Resolved by only adding the `temp://` source if the whole process succeeds. If any part fails, an appropriate `error!` message will be logged.